feat: implemented loading and error state handling

jayveeannsoriano opened the pull request:
- created and added a general action for loading state
- created an async model for the 3 app state: data, loading, and error
- wrapped List<Pokemon> in Async to utilize the newly added app state model
- modified pokemon_overview_page.dart
- added a progress/wait state in AppState model
- added a guard claus in pokemon_overview_vm. dart for the newly added app state model
- added a key in GetPokemonsAction for actionKey constructor in LoadingAction

- app state from error to working app
https://user-images.githubusercontent.com/111727585/227859170-a6f32b8f-63bb-426c-8bf1-dd42acc5ef05.mp4

- modified app error messages
https://user-images.githubusercontent.com/111727585/227886974-49d71d0d-2642-4185-b031-6ffcc4bfff25.mp4




- flutter lint analyzer results and format
![image](https://user-images.githubusercontent.com/111727585/227858976-8b5d9a33-5e8c-4670-9629-97497545e3ef.png)

Commit:
feat: implemented loading and error state handling
- created and added a general action for loading state
- created an async model for the 3 app state: data, loading, and error
- wrapped List<Pokemon> in Async to utilize the newly added app state model
- modified pokemon_overview_page.dart
- added a progress/wait state in AppState model
- added a guard claus in pokemon_overview_vm. dart for the newly added app state model
- added a key in GetPokemonsAction for actionKey constructor in LoadingAction

Commit:
feat: implemented loading and error state handling
- added a snackbar error message
- added a final variable in constants.dart for ScaffoldMessengerState key
- added a documentation comment for LoadingAction

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -9,20 +10,32 @@ class PokemonOverviewPage extends StatelessWidget {
+          ),
+          loading: () => const Center(child: CircularProgressIndicator()),
+          error: (errorMessage) {
+            final SnackBar snackBar = SnackBar(
```
kindly wrap this snackbar implementation in a separate function and name it with `_showErrorMessageSnackbar` 🙂

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -9,20 +10,32 @@ class PokemonOverviewPage extends StatelessWidget {
+          error: (errorMessage) {
+            final SnackBar snackBar = SnackBar(
+              content: Text(errorMessage!),
+              duration: const Duration(seconds: 5),
```
no need for this, just rely for its default delay value 🙂

shawn-ffufm commented on lib/utils/constants.dart:
```diff
@@ -15,3 +15,7 @@ const double imageHeight = 130.0;
     'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/[index].png';
+
+const String errorMessage = 'Something went wrong!';
+const String emptyListErrorMessage = 'No Pokemons Available';
```
this is not an error message right? instead rename this to `emptyPokemonsLabel`

shawn-ffufm commented on lib/utils/constants.dart:
```diff
@@ -15,3 +15,7 @@ const double imageHeight = 130.0;
+
+const String errorMessage = 'Something went wrong!';
+const String emptyListErrorMessage = 'No Pokemons Available';
+final scaffoldKey = GlobalKey<ScaffoldMessengerState>();
```
no need for this, just directly place this as a constant above ur `PokemonOverviewPage` but don't forget to make it private 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

Commit:
feat: implemented loading and error state handling
- made snackbar implementation a separate private function
- removed snackbar time duration declaration
- renamed emptyListErrorMessage variable name to emptyPokemonsLabel
- made scaffoldKey a private variable in PokemonOverviewPage
- removed unused import in constants.dart

Commit:
feat: implemented loading and error state handling
- removed _scaffoldKey and utilized ScaffoldMessenger
- modified _showErrorMessageSnackbar function

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -9,23 +10,38 @@ class PokemonOverviewPage extends StatelessWidget {
-            final pokemon = pokemons[index];
-            return PokemonTileCard(pokemon: pokemon);
-          },
+      home: Builder(
```
why wrap this in a Builder? seems no need for this u can maximize using the build method's provided context 🙂

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -9,23 +10,38 @@ class PokemonOverviewPage extends StatelessWidget {
   }
+
+  _showErrorMessageSnackbar(BuildContext context, String? errorMessage) {
+    ScaffoldMessenger.of(context).showSnackBar(SnackBar(content: Text(errorMessage!)));
```
use null coalescing and use `emptyString` constant as its fallback value instead of using non-null enforce here to be safer 🙂

also, don't forget to apply trailing comma here 🙂

shawn-ffufm reviewed (changes requested):
see minor pr feedback 🙂

Commit:
feat: implemented loading and error state handling
(pr feedback)
- modified main.dart to wrap pages in MaterialApp widget
- removed wrapped Build in Scaffold
- used null coalescing and emptyString constant
- added trailing commas
- used fat arrow notation in WidgetsBinding.instance...
